xdsl/dialects/func.py

```diff
@@ -47,6 +56,97 @@ def verify_(self) -> None:
                 "input types"
             )
 
+    @classmethod
+    def parse(cls, parser: Parser) -> FuncOp:
+        # Parse visibility keyword if present
+        visibility = parser.tokenizer.next_token_of_pattern(
+            re.compile("public|nested|private")
+        )
+        if isinstance(visibility, Span):
+            visibility = visibility.text
+
+        # Parse function name
+        name = parser.parse_symbol_name().data
+
+        # Parse function arguments
+        args = parser.parse_comma_separated_list(
+            parser.Delimiter.PAREN,
+            lambda: parser.parse_optional_argument()
+            or parser.parse_optional_type()
+            or parser.raise_error("Expected argument or type"),
+        )
+
+        # Check consistency (They should be either all named or none)
+        if isa(args, list[parser.Argument]):
+            entry_args = args
+            input_types = [a.type for a in args]
+            assert isa(input_types, list[Attribute])
+        elif isa(args, list[Attribute]):
+            entry_args = None
+            input_types = args
+        else:
+            parser.raise_error(
+                "Expected all arguments to be named or all arguments to be unnamed."
+            )
+
+        # Parse return type
+        if parser.parse_optional_punctuation("->"):
+            return_types = parser.parse_optional_type()
+            if return_types:
+                return_types = [return_types]
+            else:
+                return_types = parser.parse_comma_separated_list(
+                    parser.Delimiter.PAREN, parser.parse_type
+                )
+        else:
+            return_types = []
+
+        attr_dict = parser.parse_optional_attr_dict_with_keyword(
+            ("sym_name", "function_type", "sym_visibility")
+        )
+
+        # Parse body
+        region = parser.parse_optional_region(entry_args) or Region()
```

Comment:
Oof, this is some python shenanigans here. I am personally against this kind of code as it makes stuff harder to read in my opinion.